packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx

```diff
@@ -0,0 +1,213 @@
+      (item: QueryBuilderParameterDragSource): void => {
+        watermarkState.setValue(item.variable.parameter);
+      },
+      [watermarkState],
+    );
+
+    const [{ isParameterValueDragOver }, dropTargetConnector] = useDrop<
+      QueryBuilderParameterDragSource,
+      void,
+      { isParameterValueDragOver: boolean }
+    >(
+      () => ({
+        accept: [QUERY_BUILDER_PARAMETER_DND_TYPE],
+        drop: (item, monitor): void => {
+          const itemType = item.variable.parameter.genericType?.value.rawType;
+          if (
+            !monitor.didDrop() &&
+            // Doing a type check, which only allows dragging and dropping parameters of the same type or of child types
+            itemType &&
+            parameterType &&
+            (isSuperType(parameterType, itemType) ||
+              parameterType.name === itemType.name)
+          ) {
+            handleDrop(item);
+          } // prevent drop event propagation to accomondate for nested DnD
+        },
+        collect: (monitor) => ({
+          isParameterValueDragOver: monitor.isOver({
+            shallow: true,
+          }),
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <>
+        {watermarkState.value && (
+          <div key={variable.name} className="panel__content__form__section">
+            <div className="panel__content__form__section__header__label">
+              {variable.name}
+            </div>
+            <div className="query-builder__parameter-editor">
+              <PanelDropZone
+                isDragOver={isParameterValueDragOver}
+                dropTargetConnector={dropTargetConnector}
+              >
+                <BasicValueSpecificationEditor
+                  valueSpecification={guaranteeNonNullable(
+                    watermarkState.value,
+                  )}
+                  setValueSpecification={(val: ValueSpecification): void => {
+                    watermarkState.setValue(val);
+                  }}
+                  graph={graph}
+                  obseverContext={
+                    watermarkState.queryBuilderState.observableContext
+                  }
+                  typeCheckOption={{
+                    expectedType: graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+                    match:
+                      parameterType ===
+                      graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+                  }}
+                  resetValue={(): void => watermarkState.setValue(undefined)}
+                />
+              </PanelDropZone>
+            </div>
+          </div>
+        )}
+      </>
+    );
+  },
+);
+
+export const QueryBuilderWatermarkEditor = observer(
+  (props: { queryBuilderState: QueryBuilderState }) => {
+    const { queryBuilderState } = props;
+
+    const watermarkState = queryBuilderState.watermarkState;
+
+    const handleClose = (): void => {
+      if (watermarkState.value instanceof PrimitiveInstanceValue) {
```

Comment:
why are we doing this logic in closing of the modal, arent we updating the value spec in real time when the user is editing the value spec editor?